projects/knora-ui/src/lib/viewer/values/color-value/color-picker/color-picker.component.ts

```diff
@@ -0,0 +1,215 @@
+import { FocusMonitor } from '@angular/cdk/a11y';
+import { coerceBooleanProperty } from '@angular/cdk/coercion';
+import { Component, DoCheck, ElementRef, HostBinding, Input, OnDestroy, Optional, Self, OnInit } from '@angular/core';
+import { ControlValueAccessor, FormBuilder, FormControl, FormGroup, FormGroupDirective, NgControl, NgForm, Validators } from '@angular/forms';
+import { ErrorStateMatcher, MatFormFieldControl, mixinErrorState, CanUpdateErrorState, CanUpdateErrorStateCtor } from '@angular/material';
+import { Subject } from 'rxjs';
+import { CustomRegex } from '../../custom-regex';
+
+/**
+ * Represents a DateTime consisting of a date and a time.
+ */
+export class ColorPicker {
+
+  /**
+   * @param color string
+   */
+  constructor(public color: string) {
+  }
+}
+
+/** Error when invalid control is dirty, touched, or submitted. */
+export class ColorPickerErrorStateMatcher implements ErrorStateMatcher {
+  isErrorState(control: FormControl | null, form: FormGroupDirective | NgForm | null): boolean {
+    const isSubmitted = form && form.submitted;
+    return !!(control && control.invalid && (control.dirty || control.touched || isSubmitted));
+  }
+}
+
+class MatInputBase {
+  constructor(
+    public _defaultErrorStateMatcher: ErrorStateMatcher,
+    public _parentForm: NgForm,
+    public _parentFormGroup: FormGroupDirective,
+    public ngControl: NgControl) { }
+}
+const _MatInputMixinBase: CanUpdateErrorStateCtor & typeof MatInputBase =
+  mixinErrorState(MatInputBase);
+
+
+@Component({
+  selector: 'kui-color-picker',
+  templateUrl: './color-picker.component.html',
+  styleUrls: ['./color-picker.component.scss'],
+  providers: [
+    { provide: MatFormFieldControl, useExisting: ColorPickerComponent }
+  ]
+})
+export class ColorPickerComponent extends _MatInputMixinBase implements OnInit, ControlValueAccessor, MatFormFieldControl<ColorPicker>, DoCheck, CanUpdateErrorState, OnDestroy {
+
+  static nextId = 0;
+
+  private _required = false;
+  private _disabled = false;
+  private _placeholder: string;
+
+  @Input() readonly = false;
+
+  @Input() errorStateMatcher: ErrorStateMatcher;
+
+  @HostBinding() id = `kui-color-picker-${ColorPickerComponent.nextId++}`;
+
+  @HostBinding('attr.aria-describedby') describedBy = '';
+
+  colorVal: string;
+  colorLabel: string;
+  colorForm: FormGroup;
+  colorValueFormControl: FormControl;
+  stateChanges = new Subject<void>();
+  focused = false;
+  errorState = false;
+  controlType = 'kui-color-picker';
+  matcher = new ColorPickerErrorStateMatcher();
+  colorValidator = [Validators.pattern(CustomRegex.COLOR_REGEX)];
```

Comment:
I think this is not needed here, but in the parent.

Reply:
When I move the custom validator to the parent, my new value is always invalid. It works fine when the child component is dealing with the custom validator. But now with all your required changes, the whole validation does not work, I get weird behaviours and I do not always understand where they come from (base comp., parent comp. or child comp.)...

Reply:
Could you commit what you have done so far so I can check?

Have you removed your implementation of `standardValidatorFunc`?

Reply:
> When I move the custom validator to the parent, my new value is always invalid. It works fine when the child component is dealing with the custom validator. But now with all your required changes, the whole validation does not work, I get weird behaviours and I do not always understand where they come from (base comp., parent comp. or child comp.)...

I tried that yesterday and had some issues because the value was an object. The value now being a string, this component should work like the int component.

Reply:
what do you mean by "The value now being a string"? do you mean it is working like the int component with one component and not parent/child component (with ControlValueAccessor and MatFormFieldControl)?